platform/darwin/src/MGLOpenGLStyleLayer.mm

```diff
@@ -0,0 +1,200 @@
+    self.mapView = mapView;
+    if (otherLayer) {
+        const mbgl::optional<std::string> belowLayerId{ otherLayer.identifier.UTF8String };
+        mapView.mbglMap->addLayer(std::move(_pendingLayer), belowLayerId);
+    } else {
+        mapView.mbglMap->addLayer(std::move(_pendingLayer));
+    }
+}
+
+- (void)removeFromMapView:(MGLMapView *)mapView {
+    auto removedLayer = mapView.mbglMap->removeLayer(self.identifier.UTF8String);
+    self.mapView = nil;
+    if (!removedLayer) {
+        return;
+    }
+    _pendingLayer = std::move(reinterpret_cast<std::unique_ptr<mbgl::style::CustomLayer> &>(removedLayer));
+    _rawLayer = _pendingLayer.get();
+}
+
+/**
+ Called immediately after a layer is added to a map view’s style.
+ 
+ This method is intended to be overridden in a subclass. You can use this method
+ to perform any setup work before the layer is used to draw a frame. For
+ example, you might use this method to compile an OpenGL shader. The default
+ implementation of this method does nothing.
+ 
+ Any resource acquired in this method must be released in
+ `-willMoveFromMapView:`.
+ 
+ @param mapView The map view to whose style the layer has been added.
+ */
+- (void)didMoveToMapView:(MGLMapView *)mapView {
+    
+}
+
+/**
+ Called immediately before a layer is removed from a map view’s style.
+ 
+ This method is intended to be overridden in a subclass. You can use this method
+ to perform any teardown work once the layer has drawn its last frame and is
+ about to be removed from the style. The default implementation of this method
+ does nothing.
+ 
+ This method may be called even if `-didMoveToMapView:` has not been called.
+ 
+ @param mapView The map view from whose style the layer is about to be removed.
+ */
+- (void)willMoveFromMapView:(MGLMapView *)mapView {
+    
+}
+
+/**
+ Called each time the layer needs to draw a new frame in a map view.
+ 
+ This method is intended to be overridden in a subclass. You can use this method
+ to draw the layer’s content. The default implementation of this method does
+ nothing.
+ 
+ Your implementation should not make any assumptions about the OpenGL state,
+ other than that the current OpenGL context is active. It may make changes to
+ the OpenGL state. It is not required to reset values such as the depth mask,
+ stencil mask, or corresponding test flags to their original values.
+ 
+ Be sure to draw your fragments with a <var>z</var> value of 1 to take advantage
+ of the opaque fragment culling, in case the style contains any opaque layers
+ above this layer.
+ 
+ @param mapView The map view to which the layer draws.
+ @param context A context structure with information defining the frame to draw.
+ */
+- (void)drawInMapView:(MGLMapView *)mapView withContext:(MGLStyleLayerDrawingContext)context {
+    
+}
+
+/**
+ Forces the map view associated with this style to redraw the receiving layer,
+ causing its drawing handler to be run.
```

Comment:
> causing its drawing handler to be run

This is a holdover from the former block-based API. I reworded this sentence in 178130267e3d79e53506549fb43c4fb2ef18e6e0.